File: iOS_RaptorHAB_xcode/RaptorHabMobile/RaptorHabMobile/MissionManager.swift
//
//  MissionManager.swift
//  RaptorHabMobile
//
//  Manages mission data, session recording, and historical data
//  Ported from Mac version for iOS
//

import Foundation
import UIKit

// MARK: - Mission Model

struct Mission: Identifiable, Codable, Hashable {
    let id: UUID
    var name: String
    let createdAt: Date
    var launchTime: Date?
    var landingTime: Date?
    var maxAltitude: Double
    var totalDistance: Double
    var burstAltitude: Double?
    var launchLocation: Coordinate?
    var landingLocation: Coordinate?
    var burstLocation: Coordinate?
    var telemetryCount: Int
    var imageCount: Int
    var notes: String
    
    struct Coordinate: Codable, Hashable {
        let latitude: Double
        let longitude: Double
        let altitude: Double
    }
    
    var folderName: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd_HH-mm"
        return "\(formatter.string(from: createdAt))_\(name.replacingOccurrences(of: " ", with: "_"))"
    }
    
    var duration: TimeInterval? {
        guard let launch = launchTime, let landing = landingTime else { return nil }
        return landing.timeIntervalSince(launch)
    }
    
    var durationFormatted: String {
        guard let dur = duration else { return "N/A" }
        let hours = Int(dur) / 3600
        let minutes = (Int(dur) % 3600) / 60
        let seconds = Int(dur) % 60
        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%d:%02d", minutes, seconds)
        }
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
    
    static func == (lhs: Mission, rhs: Mission) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - Recorded Image

struct RecordedImage: Identifiable, Codable {
    let id: UUID
    let imageId: UInt16
    let timestamp: Date
    let filename: String
    var latitude: Double?
    var longitude: Double?
    var altitude: Double?
}

// MARK: - Mission Manager

class MissionManager: ObservableObject {
    static let shared = MissionManager()
    
    // Current session
    @Published var isRecording = false
    @Published var isAutoRecording = false
    @Published var currentMission: Mission?
    @Published var recordedTelemetry: [TelemetryPoint] = []
    @Published var recordedImages: [RecordedImage] = []
    
    // Settings
    @Published var autoRecord: Bool {
        didSet { UserDefaults.standard.set(autoRecord, forKey: "MissionAutoRecord") }
    }
    
    // Historical missions
    @Published var missions: [Mission] = []
    
    // Playback state
    @Published var isPlaying = false
    @Published var playbackMission: Mission?
    @Published var playbackTelemetry: [TelemetryPoint] = []
    @Published var playbackIndex: Int = 0
    @Published var playbackSpeed: Double = 1.0
    
    private var playbackTimer: Timer?
    private let missionsFolder: URL
    
    var hasUnsavedRecording: Bool {
        isRecording && !recordedTelemetry.isEmpty
    }
    
    private init() {
        // Setup missions folder in Documents
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        missionsFolder = documents.appendingPathComponent("RaptorHabMobile/Missions")
        try? FileManager.default.createDirectory(at: missionsFolder, withIntermediateDirectories: true)
        
        // Load settings
        autoRecord = UserDefaults.standard.bool(forKey: "MissionAutoRecord")
        
        // Load missions
        loadMissions()
    }
    
    // MARK: - Recording Control
    
    func startRecording(name: String = "Mission", isAuto: Bool = false) {
        guard !isRecording else { return }
        
        let mission = Mission(
            id: UUID(),
            name: name,
            createdAt: Date(),
            launchTime: nil,
            landingTime: nil,
            maxAltitude: 0,
            totalDistance: 0,
            burstAltitude: nil,
            launchLocation: nil,
            landingLocation: nil,
            burstLocation: nil,
            telemetryCount: 0,
            imageCount: 0,
            notes: ""
        )
        
        currentMission = mission
        recordedTelemetry.removeAll()
        recordedImages.removeAll()
        isRecording = true
        isAutoRecording = isAuto
        
        // Create mission folder
        let missionFolder = missionsFolder.appendingPathComponent(mission.folderName)
        try? FileManager.default.createDirectory(at: missionFolder, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: missionFolder.appendingPathComponent("images"), withIntermediateDirectories: true)
    }
    
    func stopRecording() {
        guard isRecording, var mission = currentMission else { return }
        
        isRecording = false
        isAutoRecording = false
        
        // Update mission stats
        mission.telemetryCount = recordedTelemetry.count
        mission.imageCount = recordedImages.count
        
        if let first = recordedTelemetry.first {
            mission.launchTime = first.timestamp
            mission.launchLocation = Mission.Coordinate(
                latitude: first.latitude,
                longitude: first.longitude,
                altitude: first.altitude
            )
        }
        
        if let last = recordedTelemetry.last {
            mission.landingTime = last.timestamp
            mission.landingLocation = Mission.Coordinate(
                latitude: last.latitude,
                longitude: last.longitude,
                altitude: last.altitude
            )
        }
        
        mission.maxAltitude = recordedTelemetry.map(\.altitude).max() ?? 0
        
        // Calculate total distance
        var distance: Double = 0
        for i in 1..<recordedTelemetry.count {
            distance += haversineDistance(
                lat1: recordedTelemetry[i-1].latitude, lon1: recordedTelemetry[i-1].longitude,
                lat2: recordedTelemetry[i].latitude, lon2: recordedTelemetry[i].longitude
            )
        }
        mission.totalDistance = distance
        
        // Get burst info if detected
        if let burst = BurstDetectionManager.shared.burstPoint {
            mission.burstAltitude = burst.altitude
            mission.burstLocation = Mission.Coordinate(
                latitude: burst.latitude,
                longitude: burst.longitude,
                altitude: burst.altitude
            )
        }
        
        currentMission = mission
        
        // Save mission data
        saveMission(mission)
        
        // Add to list
        missions.insert(mission, at: 0)
    }
    
    func discardRecording() {
        guard isRecording, let mission = currentMission else { return }
        
        // Delete the mission folder
        let missionFolder = missionsFolder.appendingPathComponent(mission.folderName)
        try? FileManager.default.removeItem(at: missionFolder)
        
        // Reset state
        isRecording = false
        isAutoRecording = false
        currentMission = nil
        recordedTelemetry.removeAll()
        recordedImages.removeAll()
    }
    
    // MARK: - Record Data
    
    func recordTelemetry(_ telemetry: TelemetryPoint) {
        // Auto-start recording if enabled and not already recording
        if autoRecord && !isRecording && telemetry.altitude > 100 {
            startRecording(name: "Auto Mission", isAuto: true)
        }
        
        guard isRecording else { return }
        
        recordedTelemetry.append(telemetry)
        
        // Auto-save periodically
        if recordedTelemetry.count % 100 == 0 {
            saveTelemetryIncremental()
        }
    }
    
    func recordImage(imageId: UInt16, data: Data, telemetry: TelemetryPoint?) {
        guard isRecording, let mission = currentMission else { return }
        
        let filename = "image_\(imageId)_\(Int(Date().timeIntervalSince1970)).jpg"
        let imagePath = missionsFolder
            .appendingPathComponent(mission.folderName)
            .appendingPathComponent("images")
            .appendingPathComponent(filename)
        
        try? data.write(to: imagePath)
        
        let recorded = RecordedImage(
            id: UUID(),
            imageId: imageId,
            timestamp: Date(),
            filename: filename,
            latitude: telemetry?.latitude,
            longitude: telemetry?.longitude,
            altitude: telemetry?.altitude
        )
        
        recordedImages.append(recorded)
    }
    
    // MARK: - Save/Load
    
    private func saveMission(_ mission: Mission) {
        let missionFolder = missionsFolder.appendingPathComponent(mission.folderName)
        
        // Save mission metadata
        let metaPath = missionFolder.appendingPathComponent("mission.json")
        if let data = try? JSONEncoder().encode(mission) {
            try? data.write(to: metaPath)
        }
        
        // Save telemetry
        let telemetryPath = missionFolder.appendingPathComponent("telemetry.json")
        if let data = try? JSONEncoder().encode(recordedTelemetry) {
            try? data.write(to: telemetryPath)
        }
        
        // Save image index
        let imagesPath = missionFolder.appendingPathComponent("images.json")
        if let data = try? JSONEncoder().encode(recordedImages) {
            try? data.write(to: imagesPath)
        }
        
        // Save telemetry CSV
        saveTelemetryCSV(missionFolder: missionFolder)
    }
    
    private func saveTelemetryIncremental() {
        guard let mission = currentMission else { return }
        let missionFolder = missionsFolder.appendingPathComponent(mission.folderName)
        let telemetryPath = missionFolder.appendingPathComponent("telemetry.json")
        if let data = try? JSONEncoder().encode(recordedTelemetry) {
            try? data.write(to: telemetryPath)
        }
    }
    
    private func saveTelemetryCSV(missionFolder: URL) {
        let csvPath = missionFolder.appendingPathComponent("telemetry.csv")
        
        var csv = "timestamp,sequence,latitude,longitude,altitude_m,speed_ms,heading,satellites,fix_type,battery_mv,rssi\n"
        
        let isoFormatter = ISO8601DateFormatter()
        
        for point in recordedTelemetry {
            let line = [
                isoFormatter.string(from: point.timestamp),
                String(point.sequence),
                String(format: "%.7f", point.latitude),
                String(format: "%.7f", point.longitude),
                String(format: "%.1f", point.altitude),
                String(format: "%.1f", point.speed),
                String(format: "%.0f", point.heading),
                String(point.satellites),
                point.fixType,
                String(point.batteryMV),
                String(point.rssi)
            ].joined(separator: ",")
            csv += line + "\n"
        }
        
        try? csv.write(to: csvPath, atomically: true, encoding: .utf8)
    }
    
    func loadMissions() {
        missions.removeAll()
        
        guard let contents = try? FileManager.default.contentsOfDirectory(
            at: missionsFolder,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: .skipsHiddenFiles
        ) else { return }
        
        for folder in contents {
            var isDir: ObjCBool = false
            guard FileManager.default.fileExists(atPath: folder.path, isDirectory: &isDir),
                  isDir.boolValue else { continue }
            
            let metaPath = folder.appendingPathComponent("mission.json")
            guard let data = try? Data(contentsOf: metaPath),
                  let mission = try? JSONDecoder().decode(Mission.self, from: data) else { continue }
            
            missions.append(mission)
        }
        
        // Sort by date (newest first)
        missions.sort { $0.createdAt > $1.createdAt }
    }
    
    // MARK: - Mission Operations
    
    func getMissionFolder(_ mission: Mission) -> URL {
        missionsFolder.appendingPathComponent(mission.folderName)
    }
    
    func loadMissionTelemetry(_ mission: Mission) -> [TelemetryPoint] {
        let path = getMissionFolder(mission).appendingPathComponent("telemetry.json")
        guard let data = try? Data(contentsOf: path),
              let telemetry = try? JSONDecoder().decode([TelemetryPoint].self, from: data) else {
            return []
        }
        return telemetry
    }
    
    func loadMissionImages(_ mission: Mission) -> [RecordedImage] {
        let path = getMissionFolder(mission).appendingPathComponent("images.json")
        guard let data = try? Data(contentsOf: path),
              let images = try? JSONDecoder().decode([RecordedImage].self, from: data) else {
            return []
        }
        return images
    }
    
    func getImageData(mission: Mission, image: RecordedImage) -> Data? {
        let imagePath = getMissionFolder(mission)
            .appendingPathComponent("images")
            .appendingPathComponent(image.filename)
        return try? Data(contentsOf: imagePath)
    }
    
    func deleteMission(_ mission: Mission) {
        let folder = getMissionFolder(mission)
        try? FileManager.default.removeItem(at: folder)
        missions.removeAll { $0.id == mission.id }
    }
    
    func updateMissionNotes(_ mission: Mission, notes: String) {
        guard let index = missions.firstIndex(where: { $0.id == mission.id }) else { return }
        
        missions[index].notes = notes
        
        // Save updated mission metadata
        let metaPath = getMissionFolder(mission).appendingPathComponent("mission.json")
        if let data = try? JSONEncoder().encode(missions[index]) {
            try? data.write(to: metaPath)
        }
    }
    
    // MARK: - Playback
    
    func startPlayback(mission: Mission) {
        stopPlayback()
        
        playbackMission = mission
        playbackTelemetry = loadMissionTelemetry(mission)
        playbackIndex = 0
        isPlaying = true
        
        guard !playbackTelemetry.isEmpty else {
            isPlaying = false
            return
        }
        
        // Calculate time between first two points to set initial interval
        var interval: TimeInterval = 1.0
        if playbackTelemetry.count >= 2 {
            interval = playbackTelemetry[1].timestamp.timeIntervalSince(playbackTelemetry[0].timestamp)
            interval = max(0.1, interval / playbackSpeed)
        }
        
        scheduleNextPlaybackPoint(interval: interval)
    }
    
    func stopPlayback() {
        playbackTimer?.invalidate()
        playbackTimer = nil
        isPlaying = false
        playbackIndex = 0
    }
    
    func pausePlayback() {
        playbackTimer?.invalidate()
        playbackTimer = nil
        isPlaying = false
    }
    
    func resumePlayback() {
        guard playbackMission != nil, playbackIndex < playbackTelemetry.count else { return }
        isPlaying = true
        scheduleNextPlaybackPoint(interval: 0.1)
    }
    
    func seekPlayback(to index: Int) {
        playbackIndex = min(max(0, index), playbackTelemetry.count - 1)
        
        // Post current telemetry
        if playbackIndex < playbackTelemetry.count {
            NotificationCenter.default.post(
                name: .playbackTelemetryUpdate,
                object: playbackTelemetry[playbackIndex]
            )
        }
    }
    
    private func scheduleNextPlaybackPoint(interval: TimeInterval) {
        playbackTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: false) { [weak self] _ in
            self?.advancePlayback()
        }
    }
    
    private func advancePlayback() {
        guard isPlaying, playbackIndex < playbackTelemetry.count else {
            stopPlayback()
            return
        }
        
        let currentPoint = playbackTelemetry[playbackIndex]
        
        // Post telemetry update
        NotificationCenter.default.post(
            name: .playbackTelemetryUpdate,
            object: currentPoint
        )
        
        playbackIndex += 1
        
        // Schedule next point
        if playbackIndex < playbackTelemetry.count {
            let nextPoint = playbackTelemetry[playbackIndex]
            var interval = nextPoint.timestamp.timeIntervalSince(currentPoint.timestamp)
            interval = max(0.05, interval / playbackSpeed)
            scheduleNextPlaybackPoint(interval: interval)
        } else {
            stopPlayback()
        }
    }
    
    // MARK: - Export
    
    func exportMissionCSV(_ mission: Mission) -> URL? {
        let telemetry = loadMissionTelemetry(mission)
        guard !telemetry.isEmpty else { return nil }
        
        let tempURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(mission.folderName).csv")
        
        var csv = "timestamp,sequence,latitude,longitude,altitude_m,speed_ms,heading,satellites,battery_mv,rssi,snr\n"
        
        let isoFormatter = ISO8601DateFormatter()
        
        for point in telemetry {
            let line = [
                isoFormatter.string(from: point.timestamp),
                String(point.sequence),
                String(format: "%.7f", point.latitude),
                String(format: "%.7f", point.longitude),
                String(format: "%.1f", point.altitude),
                String(format: "%.1f", point.speed),
                String(format: "%.0f", point.heading),
                String(point.satellites),
                String(point.batteryMV),
                String(point.rssi),
                String(format: "%.1f", point.snr)
            ].joined(separator: ",")
            csv += line + "\n"
        }
        
        try? csv.write(to: tempURL, atomically: true, encoding: .utf8)
        return tempURL
    }
    
    // MARK: - Helpers
    
    private func haversineDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double) -> Double {
        let R = 6371000.0
        let dLat = (lat2 - lat1) * .pi / 180
        let dLon = (lon2 - lon1) * .pi / 180
        let a = sin(dLat/2) * sin(dLat/2) +
                cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) *
                sin(dLon/2) * sin(dLon/2)
        let c = 2 * atan2(sqrt(a), sqrt(1-a))
        return R * c
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let playbackTelemetryUpdate = Notification.Name("playbackTelemetryUpdate")
}

File: RaptorHABGS_Python/raptorhabgs/core/mission_manager.py
"""
Mission recording and playback manager.
Automatically records telemetry, images, and flight data.
"""

import json
import csv
import shutil
from datetime import datetime
from pathlib import Path
from typing import Optional, List
from dataclasses import dataclass, field, asdict

from .telemetry import TelemetryPoint, Mission
from .config import get_data_directory


class MissionManager:
    """
    Manages mission recording and playback.
    
    Features:
    - Auto-record on first telemetry
    - Save telemetry to CSV and JSON
    - Track mission statistics
    - Copy images to mission folder
    """
    
    def __init__(self):
        self.current_mission: Optional[Mission] = None
        self.is_recording: bool = False
        self.is_auto_recording: bool = False
        
        # Recorded data
        self.recorded_telemetry: List[TelemetryPoint] = []
        self.recorded_images: List[str] = []
        
        # Mission folder
        self.mission_folder: Optional[Path] = None
        
        # Statistics tracking
        self.max_altitude: float = 0.0
        self.total_distance: float = 0.0
        self.last_position: Optional[tuple] = None
        
        # Auto-record settings
        self.auto_record_enabled: bool = True
        self.auto_record_triggered: bool = False
    
    def start_recording(self, name: str = None, auto: bool = False) -> bool:
        """
        Start recording a new mission.
        
        Args:
            name: Mission name (auto-generated if not provided)
            auto: Whether this is auto-triggered recording
        
        Returns:
            True if recording started successfully
        """
        if self.is_recording:
            return False
        
        # Create mission
        self.current_mission = Mission()
        
        if name:
            self.current_mission.name = name
        else:
            self.current_mission.name = datetime.now().strftime("Mission_%Y%m%d_%H%M%S")
        
        # Create mission folder
        missions_dir = get_data_directory() / "missions"
        self.mission_folder = missions_dir / self.current_mission.folder_name
        self.mission_folder.mkdir(parents=True, exist_ok=True)
        
        # Create subdirectories
        (self.mission_folder / "images").mkdir(exist_ok=True)
        (self.mission_folder / "telemetry").mkdir(exist_ok=True)
        
        # Reset state
        self.recorded_telemetry.clear()
        self.recorded_images.clear()
        self.max_altitude = 0.0
        self.total_distance = 0.0
        self.last_position = None
        
        self.is_recording = True
        self.is_auto_recording = auto
        
        # Save initial mission metadata
        self._save_mission_metadata()
        
        return True
    
    def stop_recording(self, save: bool = True) -> Optional[str]:
        """
        Stop recording and optionally save the mission.
        
        Args:
            save: Whether to save the mission data
        
        Returns:
            Path to mission folder if saved, None otherwise
        """
        if not self.is_recording:
            return None
        
        self.is_recording = False
        
        if save and self.current_mission and self.mission_folder:
            # Update mission stats
            self.current_mission.landing_time = datetime.now()
            self.current_mission.max_altitude = self.max_altitude
            self.current_mission.total_distance = self.total_distance
            self.current_mission.telemetry_count = len(self.recorded_telemetry)
            self.current_mission.image_count = len(self.recorded_images)
            
            # Save final data
            self._save_mission_metadata()
            self._save_telemetry_csv()
            self._save_telemetry_json()
            
            folder = str(self.mission_folder)
            
            # Reset
            self.current_mission = None
            self.mission_folder = None
            self.is_auto_recording = False
            
            return folder
        else:
            # Discard mission
            if self.mission_folder and self.mission_folder.exists():
                shutil.rmtree(self.mission_folder)
            
            self.current_mission = None
            self.mission_folder = None
            self.is_auto_recording = False
            
            return None
    
    def record_telemetry(self, telemetry: TelemetryPoint):
        """Record a telemetry point."""
        # Auto-start recording on first valid telemetry
        if self.auto_record_enabled and not self.is_recording and not self.auto_record_triggered:
            if telemetry.is_valid and telemetry.altitude > 50:
                self.auto_record_triggered = True
                self.start_recording(auto=True)
        
        if not self.is_recording:
            return
        
        self.recorded_telemetry.append(telemetry)
        
        # Update statistics
        if telemetry.altitude > self.max_altitude:
            self.max_altitude = telemetry.altitude
        
        # Track distance
        if telemetry.is_valid:
            current_pos = (telemetry.latitude, telemetry.longitude)
            if self.last_position:
                dist = self._haversine(
                    self.last_position[0], self.last_position[1],
                    current_pos[0], current_pos[1]
                )
                self.total_distance += dist
            self.last_position = current_pos
        
        # Set launch time on first telemetry
        if self.current_mission and not self.current_mission.launch_time:
            if telemetry.altitude > 100 and telemetry.vertical_speed > 1:
                self.current_mission.launch_time = datetime.now()
        
        # Update mission stats
        if self.current_mission:
            self.current_mission.max_altitude = self.max_altitude
            self.current_mission.telemetry_count = len(self.recorded_telemetry)
    
    def record_image(self, image_path: str, image_id: int):
        """Record an image to the mission."""
        if not self.is_recording or not self.mission_folder:
            return
        
        self.recorded_images.append(image_path)
        
        # Copy image to mission folder
        src = Path(image_path)
        if src.exists():
            dst = self.mission_folder / "images" / src.name
            shutil.copy2(src, dst)
        
        if self.current_mission:
            self.current_mission.image_count = len(self.recorded_images)
    
    def _save_mission_metadata(self):
        """Save mission metadata to JSON."""
        if not self.current_mission or not self.mission_folder:
            return
        
        metadata_path = self.mission_folder / "mission.json"
        with open(metadata_path, 'w') as f:
            json.dump(self.current_mission.to_dict(), f, indent=2)
    
    def _save_telemetry_csv(self):
        """Save telemetry to CSV."""
        if not self.recorded_telemetry or not self.mission_folder:
            return
        
        csv_path = self.mission_folder / "telemetry" / "telemetry.csv"
        
        with open(csv_path, 'w', newline='') as f:
            writer = csv.writer(f)
            
            # Header
            writer.writerow([
                'timestamp', 'sequence', 'latitude', 'longitude', 'altitude_m',
                'speed_ms', 'heading', 'vertical_speed_ms', 'satellites',
                'battery_mv', 'cpu_temp_c', 'rssi', 'snr'
            ])
            
            # Data
            for t in self.recorded_telemetry:
                writer.writerow([
                    t.timestamp.isoformat(),
                    t.sequence,
                    t.latitude,
                    t.longitude,
                    t.altitude,
                    t.speed,
                    t.heading,
                    t.vertical_speed,
                    t.satellites,
                    t.battery_mv,
                    t.cpu_temp,
                    t.rx_rssi,
                    t.rx_snr
                ])
    
    def _save_telemetry_json(self):
        """Save telemetry to JSON."""
        if not self.recorded_telemetry or not self.mission_folder:
            return
        
        json_path = self.mission_folder / "telemetry" / "telemetry.json"
        
        data = [t.to_dict() for t in self.recorded_telemetry]
        
        with open(json_path, 'w') as f:
            json.dump(data, f, indent=2)
    
    @staticmethod
    def list_missions() -> List[Mission]:
        """List all recorded missions."""
        missions = []
        missions_dir = get_data_directory() / "missions"
        
        if not missions_dir.exists():
            return missions
        
        for folder in sorted(missions_dir.iterdir(), reverse=True):
            if not folder.is_dir():
                continue
            
            metadata_path = folder / "mission.json"
            if metadata_path.exists():
                try:
                    with open(metadata_path) as f:
                        data = json.load(f)
                    mission = Mission.from_dict(data)
                    missions.append(mission)
                except Exception:
                    pass
        
        return missions
    
    @staticmethod
    def load_mission(mission_id: str) -> Optional[dict]:
        """Load a mission's full data."""
        missions_dir = get_data_directory() / "missions"
        
        for folder in missions_dir.iterdir():
            metadata_path = folder / "mission.json"
            if metadata_path.exists():
                try:
                    with open(metadata_path) as f:
                        data = json.load(f)
                    if data.get('id') == mission_id:
                        # Load telemetry
                        telemetry_path = folder / "telemetry" / "telemetry.json"
                        if telemetry_path.exists():
                            with open(telemetry_path) as f:
                                data['telemetry'] = json.load(f)
                        
                        # List images
                        images_dir = folder / "images"
                        if images_dir.exists():
                            data['images'] = [str(p) for p in images_dir.glob("*.webp")]
                        
                        data['folder'] = str(folder)
                        return data
                except Exception:
                    pass
        
        return None
    
    @staticmethod
    def delete_mission(mission_id: str) -> bool:
        """Delete a mission."""
        missions_dir = get_data_directory() / "missions"
        
        for folder in missions_dir.iterdir():
            metadata_path = folder / "mission.json"
            if metadata_path.exists():
                try:
                    with open(metadata_path) as f:
                        data = json.load(f)
                    if data.get('id') == mission_id:
                        shutil.rmtree(folder)
                        return True
                except Exception:
                    pass
        
        return False
    
    @staticmethod
    def _haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """Calculate distance between two points in meters."""
        import math
        R = 6371000
        
        phi1 = math.radians(lat1)
        phi2 = math.radians(lat2)
        dphi = math.radians(lat2 - lat1)
        dlambda = math.radians(lon2 - lon1)
        
        a = math.sin(dphi/2)**2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlambda/2)**2
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
        
        return R * c
    
    @property
    def has_unsaved_recording(self) -> bool:
        """Check if there's an unsaved recording."""
        return self.is_recording and len(self.recorded_telemetry) > 0
